### ISSUE TITLE
"On start of Layout" Events - Slower on C3 Runtime

### ISSUE THREAD

**Issue by joelmayer87** [Scirra/Construct-3-bugs#2519]
## Problem description

I have a problem since porting my project over to the new runtime. For some reason, "On start of Layout" events are triggered significantly slower than with the old runtime. So much so that it becomes noticable. To be fair, i unfortunately wasn't able to reproduce it with a minimal file but my project runs the exact same events but on C3 you can tell that the "on start..." events are slower.

The best way to demonstrate this is with the Fade In function i made which does nothing else than creating a black sprite and scaling it over the width and height of the screen at the beginning of a layout. It has a Fade Behaviour attached so it will fade out thus simulating a fade in effect at the start of scenes.

## Attach a .c3p

UPDATE:

I was able to replicate it after all! Please see this c3p file:

[onstartoflayout_bug.zip](https://github.com/Scirra/Construct-3-bugs/files/2768311/onstartoflayout_bug.zip)

Note: If you have trouble seeing the effect try making the Fade Sprite a black color instead of blue.

![image](https://user-images.githubusercontent.com/30116964/51269612-2b764880-19c3-11e9-8f5f-d38f966cd269.png)
Here's the function in question.

![fadein_noerror](https://user-images.githubusercontent.com/30116964/51269743-8e67df80-19c3-11e9-9dd4-1f0a33d627ef.gif)
Here is how it worked thus far in the C2 runtime. Never gave me problems.

![fadein_error](https://user-images.githubusercontent.com/30116964/51269863-d4bd3e80-19c3-11e9-99a8-32a9c6826d0d.gif)
And here's the exact same scene and exact same code in the C3 runtime. You can see, before the black screen is called the scene is visible for a split-second. So the creation of the Fade-In object is either slower or something is delaying it. I didn't change the function or anything though and it used to work perfectly before.

Again i'm sorry i can't reproduce it with a smaller file. In case you want to look at my project, i can prepare it for you in a way that you see what's going on right away hopefully and send it over.

##Steps to reproduce

1. Open attached file in Construct
2. Start Layout with C3 runtime
3. Fade object loads too slowly (maybe press CTRL+R a couple of times to see the effect)
4. Switch to C2 runtime
5. Fade object loads as expected.

## Observed result

Delayed loading of objects on "On Start of Layout" events compared to C2.

## Expected result

Same results as in C2.

## First affected release

r134

## Affected browsers/platforms

* Chrome: yes

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 71.0.3578.98
Browser engine: Blink
Browser architecture: (unknown)
Context: browser
Operating system: Mac OS X
Operating system version: 10.14_0
Operating system architecture: (unknown)
Device type: desktop
Device pixel ratio: 2
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36
C3 release: r134 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 132 gb
Storage usage (approx): 58 mb (0%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: ATI Technologies Inc.
Renderer: AMD Radeon R9 M395X OpenGL Engine
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 2048
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by joelmayer87** [comment]
I was able to replicate the problem with a small file and attached it to the report (see above).

**Comment by ChadoriRebornXD** [comment]
[onstartoflayout_fixed.zip](https://github.com/Scirra/Construct-3-bugs/files/2768336/onstartoflayout_fixed.zip)

**Here is the fixed version**. It seems that it's not actually a **Start of Layout** bug. It's just that the **Loader Layout On Complete** is not yet finished on C3Runtime.
Just wanted to make things clear.
From my standpoint, **C2Runtime** seems to be loading it on **lazy** and **C3Runtime** loading it all on one-shot.  That's just my guess though from a quick glance.

**Comment by joelmayer87** [comment]
I implemented Chadori's solution to the main game and while it seems to be working with the small example project, it doesn't make any difference to the main project and the flashing and late loading still occurs even if i go to a loader layout first and call the scene after that.

**Comment by AshleyScirra** [comment]
I can't see what's wrong with the minimal repro. It seems to fade in immediately with both runtimes. It's not clear to me why you have two layouts in the project though. Does it reproduce with one, but not the other? If so which is which? The steps don't seem to include which to test, or whether to preview project vs. preview layout.

Maybe there's some issue where in some cases it can draw the layout before 'On start of layout' fires. The best way to demonstrate this would be a red background and putting a different color object on top of it in 'On start of layout'. Then if you can ever see red, it's a bug. Well, I tried that: 
[startoflayout-sequence.zip](https://github.com/Scirra/Construct-3-bugs/files/2772992/startoflayout-sequence.zip) - and it never shows red. So as far as I can tell it's always triggering 'On start of layout' before drawing the screen.

I think I need a better repro project to be able to investigate this further.

**Comment by SnipG** [comment]
Its bugged when created object has no instances on layout. 
[BugSpawn.zip](https://github.com/Scirra/Construct-3-bugs/files/2773232/BugSpawn.zip)

There is other bug too, that newly created instance image point is messed up somehow. And system will put X Y is wrong for that new instance. --> who has time can make new issue report for this.

**Comment by joelmayer87** [comment]
![objectsnotonlayout](https://user-images.githubusercontent.com/30116964/51393686-95642e80-1b38-11e9-8327-2d788c5325d9.gif)

Here you can see it really clearly. Since it's the first scene of my game i hide the portrait and the cell phone in my inventory by simply pasting a piece of inventory on top of it. Plus of course the usual fade object. In gif they are created from the Bank layout. So you can see them flashing on loading the Layout.

![objectsonlayout](https://user-images.githubusercontent.com/30116964/51393746-b9277480-1b38-11e9-872c-d5c42c65bfd3.gif)

And here i moved them on the same layout. Everything loads as expected.

**Comment by SnipG** [comment]
You should not use sprite banks etc. Its bad practice.
 I know why people use it. But for scirra to make sprite management simpler, they could add in special sprite canvas: On top of current or outside viewport, where you can scroll with special button etc.

Anyway nice you got it fixed for youself.

**Comment by joelmayer87** [comment]
Well i mean i was just following like every tutorial out there and it worked perfectly on the C2 runtime :D I'd still like Scirra to look into this because it's unexpected behaviour compared to the C2 runtime where it still loads normally no matter where the bank is. I know though, that we don't need Sprite Banks anymore with the new runtime but this project was ported from the old runtime. Thanks again! :)

**Comment by AshleyScirra** [comment]
This is by design. It's a conscious design change we made in the C3 runtime.

If the object is not placed on the layout, Construct interprets that as meaning the object is not used on that layout. Our advice for memory management to work best is that all objects that are used on the layout are placed somewhere on it, and then they're destroyed on Start of Layout if they're not immediately needed. This then means Construct sees the right sets of objects to load images for, and can start up the layout efficiently by ensuring it's pre-loaded all the right images by the time it starts it.

If an object is not placed on the layout but you create it with events, it's not loaded in to memory. This forces Construct to do an on-the-spot memory load for the sprite's images. In Construct 2 it did this synchronously, which janked the game (and everyone hates jank). It also precludes the possibility of loading multiple images in parallel, which makes it slower if it has to load multiple images (and slows down loading layouts, since they load images synchronously too). Construct 3 does this asynchronously and in parallel, so it doesn't jank the game, and can load multiple images in parallel, which can be much faster.

The downside is it creates the situation where it has to draw objects that don't have images loaded yet, because the object was created and the game carried on running while the images load. The C3 runtime just skips drawing the object in this case - so they are essentially invisible until their images load. That's what you're seeing in this case: there's nothing wrong with 'On start of layout', it's just showing the object invisible for a couple of frames until its images load.

The solution: use memory management correctly - make sure all objects that the layout uses are placed on the layout. Then everything's loaded by 'On start of layout' and it doesn't have to do on-the-spot loads. This is more efficient, helps Construct manage memory more efficiently, and as of recently, also helps Construct arrange spritesheets efficiently based on the objects layouts use to save memory. So not only does this solve the problem, it's a really good idea to do it anyway!

**Comment by joelmayer87** [comment]
Ok thank you for clearing that up!
